submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,293 @@
+from the backend Pods will have the source as the globalIP of the exportedService.
+
+On the other hand, when the backend Pod tries to talk to the external application in remote
+Cluster, the external application will see that traffic is coming with a globalIP that is
+assigned either at the Cluster level, or namespace level, or the one requested by the user
+at Pod level. This asynchronous behavior will be analogous to the behavior seen with standard
+K8s applications when they talk to any Services. Applications desiring to have the same IPaddress
+for both ingress and egress traffic can create Headless Services in front of them and should
+avoid requesting explicit globalIP that match Headless Service backend Pods.
+
+The proposal has its own set of Pros and Cons.
+
+### Pros
+
+Since we will be annotating only select K8s objects with globalIP, it will help in improved scalability.
+Also, this translates to fewer iptable rules on the active Gateway node of the cluster which will
+improve the overall performance of the solution.
+
+### Cons
+
+A limitation with the proposed approach is that we will lose the unique identity for **every**
+Pod which is supported today. For users that intend to use Network Policies across Clusters or
+for debugging purposes, this can be mitigated by creating the necessary GlobalnetEgressIP CRDs
+which create unique globalIPs at the Namespace/Pod level.
+
+### Caveats
+
+In the current implementation, when a client Pod is scheduled, until it's annotated with a globalIP,
+and the corresponding egress rules are programmed on the node, it was not able to communicate with
+remote Services. This was creating some issues in-terms of user-experience. In this proposal, as we
+have a globalIP at the cluster level, a Pod can communicate with remote Services right away which
+will be closer to the Vanilla Submariner experience.
+
+## Design Details
+
+The following CRD is proposed to support the new design.
+
+```Go
+type GlobalnetEgressIP struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec is the specification of desired behavior of GlobalnetEgressIP object.
+    Spec               EgressIPSpec `json:"spec"`
+
+    // Observed status of GlobalnetEgressIP. Its a read-only field.
+    // +optional
+    Status             EgressIPStatus `json:"status,omitempty"`
+}
+
+type EgressIPSpec struct {
+    // GlobalIPs is a list of EgressIPs requested. The requested globalIPs should be part
+    // of globalCIDR of the Cluster. If the requested globalIPs are not allocated to other
+    // K8s objects, then it will be allocated and the status field will be updated accordingly.
+    // If the globalIPs are already allocated to other objects, creation of the CRD will fail.
+    // If this field is not specified, a single globalIP will be allocated.
+    // This field cannot be changed through updates.
+    // +optional
+    GlobalIPs  []string `json:"globalIPs,omitempty"`
```

Comment:
Can we do this?, or should we have GlobalIPs CIDR?

Reply:
What if we have 

     GlobalIPCIDR string ...    (optional)

Which could be used like
 ```
     GlobalIPCIDR: 169.254.4.1/32  or 169.254.4.1
     GlobalIPCIDR: 169.254.4.0/29
```
or
```
    GlobalIPCIDR: "/29"   meaning that you want a /29 and you don't care about the 
```

Reply:
IMHO, let's keep it simple, the more options we provide the more corner/edge cases that needs to be handled. 

Since these are EgressIPs, from an implementation POV, if we take CIDR input, we still have to specify individual IPTable SNAT rules (AFAICT IPTables does not support specifying a CIDR for SNAT target).

On a related note, OpenShift EgressIP feature does not support CIDRs, it only supports an array. I'm trying to keep our GlobalnetEgressIP feature a bit aligned to the EgressIP use-case from OCP.

In most of the cases, users will just need a single GlobalIP. The array is to provide them flexibility to specify multiple GlobalIPs. If we really see a need for CIDRs, we can do it as future enhancement once the implementation matures.

Reply:
Two distinct fields would be easier I guess. What is your opinion on the following 

```
// User can explicitly request a specific GlobalIP and this should be a
// valid IPaddress from globalnetCIDR allocated to the Cluster. 
// If the requested globalIP falls outside of the globalnetCIDR or is
// already allocated, then Status field will be set to Error with
// appropriate message.
// If this field is not specified, a single globalIP will be allocated.
// This field cannot be changed through updates.
// +optional
GlobalIP      string  `json:"globalIP,omitempty"`

// The number of globalIPs requested.
// User can specify this field while omitting GlobalIP. In such cases,
// Globalnet will allocate the requested number of contiguous GlobalIPs
// for this GlobalnetEgressIP object.
// +optional
NumGlobalIPs  int     'json:"numGlobalIPs",omitempty"`
```

Reply:
Yeah it looks good.

Not sure why the fields shouldn't be able to change though? Perhaps initially it makes sense, but generally what's wrong with changing the global IP?

Reply:
OK sure just wanted to make sure I'm understanding correctly. It sounded like a prioritization issue to me, rather than a real limitation.